### DISCUSSION TITLE
How to replace a Identifier node with a Literal node?

### DISCUSSION THREAD

**Discussion by rattlecattle** [oxc-project/oxc#8065] [Q&A] [answered]
I'm building a constant folding optimization pass. This is a contrived example of what I'm trying to achieve.

This is the orignal JS code.

```JS
const PI = 3.14;

function getPI() {
  return PI;
}

function area(r) {
  return PI * r * r;
}
```

I want to simplify it to the below.
```JS

function getPI() {
  return 3.14;
}

function area(r) {
  return 3.14 * r * r;
}
```

Using Scope and Bindings can already find out the places the variable `PI` is referenced. My question is how to replace occurences of the Identifier `PI` with the Literal `3.14` without implementing `visit_xxx` for `VisitMut` for every type of parent node the Identifier may be in?

For example,
In `getPI`, the parent of `PI` is `ReturnStatement`. 
In `area` the parent is `BinaryExpression`.

Is there a way to avoid implemeting `visit_xxx` for each of `ReturnStatement` , `BinaryExpression` and so on to check if the child is an identifier and replace it with a Literal?

**Comment by overlookmotel** [comment, marked as answer]
`PI` is an `IdentifierReference`, which is a variant of `Expression` enum. So you can use `VisitMut::visit_expression` and replace that `Expression` with a *different kind* of `Expression`.

```rs
use oxc_ast::{
    ast::{Expression, NumberBase},
    visit::VisitMut,
    AstBuilder,
};

struct PiReplacer<'a> {
    ast_builder: AstBuilder<'a>,
}

impl<'a> VisitMut<'a> for PiReplacer<'a> {
    fn visit_expression(&mut self, expr: &mut Expression<'a>) {
        if let Expression::Identifier(ident) = expr {
            if ident.name == "PI" {
                *expr = self.ast_builder.expression_numeric_literal(
                    ident.span,
                    3.14,
                    None,
                    NumberBase::Decimal,
                );
            }
        }
    }
}
```
